packages/eslint-plugin/src/rules/no-base-to-string.ts

```diff
@@ -154,39 +161,57 @@ export default createRule<Options, MessageIds>({
       return Usefulness.Always;
     }
 
-    function collectArrayCertainty(type: ts.Type): Usefulness {
+    function collectArrayCertainty(
+      type: ts.Type,
+      visited: Set<ts.Type>,
+    ): Usefulness {
       const elemType = nullThrows(
         type.getNumberIndexType(),
         'array should have number index type',
       );
-      return collectToStringCertainty(elemType);
+      return collectToStringCertainty(elemType, visited);
     }
 
-    function collectJoinCertainty(type: ts.Type): Usefulness {
+    function collectJoinCertainty(
+      type: ts.Type,
+      visited: Set<ts.Type>,
+    ): Usefulness {
       if (tsutils.isUnionType(type)) {
```

Comment:
I haven't been able to come up with a test case, but should this have a short-circuit as well?
```suggestion
      if (visited.has(type) {
        return Usefulness.Always;
      }
      
      if (tsutils.isUnionType(type)) {
```

Reply:
Good point!

I thought about this, too, and I don't think so, but I'm not 100% sure about this.

To my understanding, a recursive type is valid if its recursion is through a box type (e.g. array, promise):

```ts
type Value = Value[];
```

A recursive type without it should show up as an error:

```ts
type Value = Value | Value[];
//   ~~~~~
// Type alias 'Value' circularly references itself.
```

Even with intersections in play, I still don't think there can be a valid recursion that doesn't use a boxed value (currently only an array).

I've tried to come up with such a case but couldn't find anything that would make the rule crash. I don't think the rule can crash without using an array type (which the current implementation covers).